README.md
# cpangitify ![linux](https://github.com/uperl/App-cpangitify/workflows/linux/badge.svg) ![macos](https://github.com/uperl/App-cpangitify/workflows/macos/badge.svg)

Convert cpan distribution from BackPAN to a git repository

# SYNOPSIS

```
% cpangitify Foo::Bar
```

# DESCRIPTION

This script fetches all known revisions of a distribution from CPAN/BackPAN
and creates a git repository with one revision and one tag for each version
of the distribution.

The idea is to create a starting point for a git work flow when adopting a
CPAN module for which you don't have access to the original repository.
It is of course better to import from Subversion or to clone an existing
git repository, but that may not be an option.

If the distribution you are migrating changed names during its history,
simply specify each name it had on the command line.  Be sure to specify
the current name first (this will be used when creating a directory name
for the repository).  For example [Mojolicious::Plugin::TtRenderer](https://metacpan.org/pod/Mojolicious::Plugin::TtRenderer) was
once called `MojoX::Renderer::TT`, so you would get both names in the
history like this:

```
% cpangitify Mojolicious::Plugin::TtRenderer MojoX::Renderer::TT
```

# OPTIONS

## --resume

Resume the import from CPAN.  Note that any local changes will be overwritten
by the CPAN upstream (your modifications will remain in the repository history).

## --output | -o _directory_

Creat the new repository in the given directory.

## --skip _version_

Skip the given versions.  Can be specified multiple times and can
be provided as a comma separated list.

## --trace

Print each git command before it is executed.

## --backpan\_index\_url

The URL to use for the BackPAN index

## --backpan\_url

The URL to use for BackPAN

## --metacpan\_url

The URL to use for metacpan.

## --branch | -b

Default branch.  As on 0.18 this is `main` by default.  Previously the old
git default was used.

## --help | -h

Print out help and exit.

## --version

Print out version and exit.

# CAVEATS

Each commit belongs to the CPAN author who submitted the corresponding release,
therefore `git blame` may not be that useful for the imported portion of
your new repository history.

The commits are ordered by date, so where there are interleaving of releases
that belong to development and production branches this simple minded script
will probably do the wrong thing.

Patches are welcome.

# SEE ALSO

Here are some similar projects:

- [Git::CPAN::Patch](https://metacpan.org/pod/Git::CPAN::Patch)

    Comes with a `git cpan import` which does something similar.  With this
    incantation I was able to get a repository for [YAML](https://metacpan.org/pod/YAML) (including history,
    but without authors and without the correct dates):

    ```
    % mkdir YAML
    % git init .
    % git cpan import --backpan YAML
    % git merge remotes/cpan/master
    ```

    One advantage here over `cpangitify` is that you should then later be able to
    import/merge future CPAN releases into yours.  [Git::CPAN::Patch](https://metacpan.org/pod/Git::CPAN::Patch) also has a bunch of
    other useful tools for creating and submitting patches and may be worth
    checking out.

    If you do an internet search for this sort of thing you may see references
    to `git-backpan-init`, but this does not appear to be part of the
    [Git::CPAN::Patch](https://metacpan.org/pod/Git::CPAN::Patch) anymore (I believe `git-import` with the `--backpan`
    option is the equivalent).

    In general `cpangitify` is a one trick poney (though good at that one thing),
    and [Git::CPAN::Patch](https://metacpan.org/pod/Git::CPAN::Patch) is a Batman's utility belt with documentation that
    (for me at least) is pretty impenetrable.

- [gitpan](https://github.com/gitpan)

    Doesn't appear to have been updated in a number of years.

- [ggoosen's cpan2git](https://github.com/ggoossen/cpan2git)

The reason I am not using the latter two is that they are designed to
mirror the whole of CPAN/BackPAN, but I'm mostly just interested in one
or two distributions here and there.

# AUTHOR

Author: Graham Ollis <plicease@cpan.org>

Contributors:

Mohammad S Anwar (MANWAR)

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013-2022 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.20      2022-09-01 19:15:57 -0600
  - Validate SSL with (gh#19)

0.19      2022-08-23 19:28:42 -0600
  - Use Archive::Libarchive::Extract instead of Archive::Extract (gh#17)
  - Drop support for Perls older than 5.20 (gh#17)

0.18      2021-04-22 02:29:50 -0600
  - Added --branch | -b option (gh#12)
  - New repositories are now created with the default branch of main (gh#12)
  - Moved to new GitHub org

0.17      2017-08-08 12:41:51 -0400
  - Update for API interface change (?) allow author email
    address as either a scalar or array reference.
  - Switch back to MakeMaker for installer

0.16      2017-07-23 06:54:53 -0400
  - Remove broken link to no longer existing module

0.15      2017-07-22 22:37:59 -0400
  - Remove dependency on File::HomeDir

0.14      2017-06-05 13:55:19 -0400
  - Use v1 metacpan API (GH#9)

0.12      2014-12-20 05:27:19 -0500
  - added --trace option
  - added --output | -o option
  - added --resume option
  - added --skip optio
  - fixes for empty commits (two sequential releases that are identical) (GH#8)
  - fixes for filenames with non ASCII characters (GH#8)
  - do not import .git directory from tarballs
  - documented some previously undocumented options

0.11      2014-10-10 11:41:51 -0400
  - 'git rm' before 'git add'
    This seems to help on MSWin32, where I may have a newer version of git than on Debian Linux

0.10      2014-08-05 13:22:41 -0400
  - use HTTP::Tiny instead of LWP
    the former being part of the core as of Perl 5.14
  - use JSON::PP instead of JSON
    the former being part of the core as of Perl 5.14
  - use CPAN::ReleaseHistory instead of BackPAN::Index
    the latter may (?) be faster,
    but the former has lighter set of prereqs

0.09      2014-06-19 16:05:18 -0400
  - CPAN release (0.08 never made it to CPAN)

0.08      2014-01-08 06:10:17 -0500
  - make multiple tags a warning only
  - install with Module::Build instead of MakeMaker
  - use Archive::Extract instead of tar
  - use recursive Path::Class::File#remove and Path::Class::Dir#remove instead of calling system rm -rf

0.07      2013-10-18 12:41:40 -0400
  - make git 1.5.0 or better a requirement

0.06      2013-10-13 15:12:49 -0400
  - documentation (mention Git::CPAN::Patch)

0.05      2013-10-13 05:39:34 -0400
  - require BackPAN::Index 0.41

0.04_01   2013-10-11 17:38:22 -0400
  - include git version in test diagnostic

0.04      2013-10-09 20:29:45 -0400
  - change name to cpangitify
  - handle distros with name changes

0.03      2013-10-09 15:43:54 -0400
  - documentation

0.02      2013-10-09 13:57:30 -0400
  - include date and author info from cpan

0.01      2013-10-09 12:58:19 -0400
  - initial version


t/00_diag.pl
use strict;
use warnings;
use Test::More;

our $format;
diag sprintf $format, 'git', eval {
  require AnyEvent::Git::Wrapper;
  AnyEvent::Git::Wrapper->new(".")->version;
} || '-';

1;


t/backpan/update.pl
use strict;
use warnings;
use v5.10;
use Path::Class qw( file dir );

my $root = file(__FILE__)->parent->subdir(qw( authors id L LO LOCAL ));
my $index = file(__FILE__)->parent->file('backpan-index.txt');
my $fh = $index->openw;

foreach my $file (sort { $a->basename cmp $b->basename } $root->children)
{
  say $fh join(' ', join('/', qw( authors id L LO LOCAL ), $file->basename), $file->stat->mtime, -s $file );
}

$root = file(__FILE__)->parent->subdir(qw( authors id P PL PLICEASE ));

foreach my $file (sort { $a->basename cmp $b->basename } $root->children)
{
  say $fh join(' ', join('/', qw( authors id P PL PLICEASE ), $file->basename), $file->stat->mtime, -s $file );
}

close $fh;



inc/Test/HTTPTinyFile.pm
package Test::HTTPTinyFile;

use strict;
use warnings;
use HTTP::Tiny;
use HTTP::Date qw( time2str );
use URI;
use Test::More ();

my $request_method = \&HTTP::Tiny::request;

BEGIN { *note = \&Test::More::note }

my $request_wrapper = sub
{
  # TODO options to support 'If-Modified-Since' see PAUSE::Packages
  my($self, $method, $url, $args) = @_;
  my $uri = URI->new($url);
  
  note "HTTP::Tiny $method $url";
  
  if($uri->scheme eq 'file')
  {
    tie my %headers, 'Test::HTTPTinyFile::ResponseHeaderTie';
    my $path = $uri->path;
    my $result = { url => $url, content => '', headers => \%headers }; # TODO include some headers
    my $content = '';
    
    if($method =~ /(GET|HEAD)/)
    {
      if(-d $path)
      {
        die "TODO";
      }
      elsif(-r $path)
      {
        if($method eq 'GET')
        {
          eval {
            use autodie;
            open my $fh, '<', $path;
            binmode $fh;
            local $/;
            $content = <$fh> if $method eq 'GET';
            close $fh;
          };
          if($@)
          {
            $result->{success} = 0;
            $result->{status} = 599;
            $result->{reason} = 'Internal Exception';
          }
        }
        unless(defined $result->{success})
        {
          $result->{success} = 1;
          $result->{status}  = 200;
          $result->{reason}  = 'OK';
          $headers{'last-modified'} = time2str((stat $path)[9]);
        }
      }
      elsif(-e $path)
      {
        $result->{success} = 0;
        $result->{status}  = 403;
        $result->{reason}  = 'Forbidden';
      }
      else
      {
        $result->{success} = 0;
        $result->{status}  = 404;
        $result->{reason}  = 'Not Found';
      }
    }
    elsif($method eq 'POST')
    {
      die "TODO";
    }
    elsif($method eq 'PUT')
    {
      die "TODO";
    }
    elsif($method eq 'DELETE')
    {
      die "TODO";
    }
    else
    {
      die "unknown HTTP method: $method";
    }
    note "HTTP::Tiny ", join(' ', $result->{success}, $result->{status}, $result->{reason});
    if($args->{data_callback})
    {
      $args->{data_callback}->($content, $result);
    }
    else
    {
      $result->{content} = $content;
    }
    return $result;
  }
  else
  {
    $request_method->(@_);
  }
};

do { no warnings; *HTTP::Tiny::request = $request_wrapper };

package
  Test::HTTPTinyFile::ResponseHeaderTie;

BEGIN { *note = \&Test::More::note }

sub TIEHASH
{
  my($class) = @_;
  bless {}, $class;
}

sub FETCH
{
  my($self, $key) = @_;
  note "header FETCH $key";
  $self->{$key};
}

sub STORE
{
  my($self, $key, $value) = @_;
  note "header STORE $key $value";
  $self->{$key} = $value;
}

sub DELETE
{
  my($self, $key) = @_;
  note "header DELETE $key";
  delete $self->{$key};
}

sub CLEAR
{
  my($self) = @_;
  note "header CLEAR";
  %$self = ();
}

sub EXISTS
{
  my($self, $key) = @_;
  note "header EXISTS $key";
  exists $self->{$key};
}

sub FIRSTKEY
{
  my($self) = @_;
  note "header FIRSTKEY";
  die "TODO";
}

sub NEXTKEY
{
  my($self, $lastkey) = @_;
  note "header NEXTKEY $lastkey";
  die "TODO";
}

sub SCALAR
{
  my($self) = @_;
  note "header SCALAR";
  scalar %$self;
}

sub DESTROY
{
  my($self) = @_;
  note "header DESTROY";
}

sub UNTIE
{
  my($self) = @_;
  note "header UNTIE";
}

1;


lib/App/cpan2git.pm
package App::cpan2git;

use strict;
use warnings;
use autodie qw( :system );
use v5.10;
use BackPAN::Index;
use Getopt::Long qw( GetOptions );
use Pod::Usage qw( pod2usage );
use Path::Class qw( file dir );
use Git::Wrapper;
use File::Temp qw( tempdir );
use LWP::UserAgent;
use File::chdir;
use JSON qw( from_json );
use URI;
use PerlX::Maybe qw( maybe );
use File::Copy::Recursive qw( rcopy );

# ABSTRACT: Convert cpan distribution from BackPAN to a git repository
# VERSION

=head1 DESCRIPTION

This is the module for the L<cpan2git> script.  See L<cpan2git> for details.

=head1 SEE ALSO

L<cpan2git>

=cut

our $ua  = LWP::UserAgent->new;
our $opt_metacpan_url;

sub main
{
  my $class = shift;
  local @ARGV = @_;
  
  my $opt_backpan_index_url;
  my $opt_backpan_url = "http://backpan.perl.org";
  $opt_metacpan_url   = "http://api.metacpan.org/";

  GetOptions(
    'backpan_index_url=s' => \$opt_backpan_index_url,
    'backpan_url=s'       => \$opt_backpan_url,
    'metacpan_url=s'      => \$opt_metacpan_url,
    'help|h'              => sub { pod2usage({ -verbose => 2}) },
    'version'             => sub {
      say 'cpan2git version ', ($App::cpan2git::VERSION // 'dev');
      exit 1;
    },
  ) || pod2usage(1);

  my @names = map { s/::/-/g; $_ } @ARGV;
  my $name = $names[0];

  pod2usage(1) unless $name;

  my $dest = dir()->absolute->subdir($name);

  if(-e $dest)
  {
    say "already exists: $dest";
    return 2;
  }

  say "creating/updating index...";
  my $bpi = BackPAN::Index->new(
    maybe backpan_index_url => $opt_backpan_index_url,
  );

  say "searching...";
  my @rel = eval { $bpi->releases->search({ dist => \@names }, { order_by => 'date' }) };

  if($@ || @rel == 0)
  {
    say "no releases found for $name";
    return 2;
  }

  say "mkdir $dest";
  $dest->mkpath(0,0700);

  my $git = Git::Wrapper->new($dest->stringify);

  $git->init;

  sub author($)
  {
    state $cache = {};
  
    my $cpanid = shift;
  
    unless(defined $cache->{$cpanid})
    {
      my $uri = URI->new($opt_metacpan_url . "v0/author/" . $cpanid);
      my $res = $ua->get($uri);
      unless($res->is_success)
      {
        say "error fetching $uri";
        say $res->status_line;
        return 2;
      }
      $cache->{$cpanid} = from_json($res->decoded_content)
    }
  
    sprintf "%s <%s>", $cache->{$cpanid}->{name}, $cache->{$cpanid}->{email}->[0];
  }

  foreach my $rel (@rel)
  {
    my $path    = $rel->path;
    my $version = $rel->version;
    my $date    = $rel->date;
    my $cpanid  = $rel->cpanid;
  
    say "$path [ $version ]";
  
    my $tmp = dir( tempdir( CLEANUP => 1 ) );
  
    local $CWD = $tmp->stringify;
  
    my $uri = URI->new(join('/', $opt_backpan_url, $path));
    say "fetch ...";
    my $res = $ua->get($uri);
    unless($res->is_success)
    {
      say "error fetching $uri";
      say $res->status_line;
      return 2;
    }
  
    do {
      open my $fh, '>', "archive";
      print $fh $res->decoded_content;
      close $fh;
    };
  
    say "unpack...";
    system 'tar', 'xf', 'archive';
    unlink 'archive';
  
    my $source = do {
      my @children = map { $_->absolute } dir()->children;
      if(@children != 1)
      {
        say "archive doesn't contain exactly one child: @children";
      }
  
      $CWD = $children[0]->stringify;
      $children[0];
    };
  
    say "merge...";
  
    foreach my $child ($dest->children)
    {
      next if $child->basename eq '.git';
      system 'rm', '-rf', $child;
    }
  
    foreach my $child ($source->children)
    {
      if(-d  $child)
      {
        rcopy($child, $dest->subdir($child->basename)) || die "unable to copy $child $!";
      }
      else
      {
        rcopy($child, $dest->file($child->basename)) || die "unable to copy $child $!";
      }
    }
  
    say "commit and tag...";
    $git->add('.');
    $git->rm($_->from) for grep { $_->mode eq 'deleted' } $git->status->get('changed');
    $git->commit({
      message => "version $version",
      date    => "$date +0000",
      author  => author $cpanid,
    });
    $git->tag($version);
  }
  
  return 0;
}

1;


lib/App/cpangitify.pm
package App::cpangitify;

use strict;
use warnings;
use autodie qw( :system );
use 5.020;
use experimental qw( signatures );
use Getopt::Long qw( GetOptions );
use Pod::Usage qw( pod2usage );
use Path::Class qw( file dir );
use Git::Wrapper;
use File::Temp qw( tempdir );
use File::chdir;
use JSON::PP qw( decode_json );
use URI;
use PerlX::Maybe qw( maybe );
use File::Copy::Recursive qw( rcopy );
use File::Basename qw( basename );
use Archive::Libarchive::Extract;
use CPAN::ReleaseHistory;
use HTTP::Tiny;

# ABSTRACT: Convert cpan distribution from BackPAN to a git repository
# VERSION

=head1 DESCRIPTION

This is the module for the L<cpangitify> script.  See L<cpangitify> for details.

=head1 SEE ALSO

L<cpangitify>

=cut

our $ua  = HTTP::Tiny->new( verify_SSL => 1 );
our $opt_metacpan_url;

sub _rm_rf ($file)
{
  if($file->is_dir && ! -l $file)
  {
    _rm_rf($_) for $file->children;
  }

  $file->remove || die "unable to delete $file";
}

our $_run_cb = sub {};
our $original_run = \&Git::Wrapper::RUN;
our $trace = 0;

sub _run_wrapper ($self, @command)
{
  my @display;
  foreach my $arg (@command)
  {
    if(ref($arg) eq 'HASH')
    {
      foreach my $k (keys %$arg)
      {
        my $v = $arg->{$k};
        push @display, "--$k";
        push @display, $v =~ /\s/ ? "'$v'" : $v
          if $v ne '1'; # yes there is a weird exception for this :P
      }
    }
    else
    {
      push @display, $arg;
    }
  }
  $_run_cb->($self, @display);
  say "+ git @display" if $trace;
  $original_run->($self, @command);
}

sub author ($cpanid)
{
  state $cache = {};

  unless(defined $cache->{$cpanid})
  {
    my $uri = URI->new($opt_metacpan_url . "v1/author/" . $cpanid);
    my $res = $ua->get($uri);
    unless($res->{success})
    {
      say "error fetching $uri";
      say $res->{reason};
      return 2;
    }
    $cache->{$cpanid} = decode_json($res->{content})
  }

  my $email = $cache->{$cpanid}->{email};
  $email = $email->[0] if ref($email) eq 'ARRAY';
  sprintf "%s <%s>", $cache->{$cpanid}->{name}, $email;
}

sub main ($, @args)
{
  local @ARGV = @args;
  no warnings 'redefine';
  local *Git::Wrapper::RUN = \&_run_wrapper;
  use warnings;

  my %skip;
  my $opt_backpan_index_url;
  my $opt_backpan_url = "http://backpan.perl.org/authors/id";
  $opt_metacpan_url   = "http://fastapi.metacpan.org/";
  my $opt_trace = 0;
  my $opt_output;
  my $opt_resume;
  my $opt_branch = 'main';

  GetOptions(
    'backpan_index_url=s' => \$opt_backpan_index_url,
    'backpan_url=s'       => \$opt_backpan_url,
    'metacpan_url=s'      => \$opt_metacpan_url,
    'trace'               => \$opt_trace,
    'skip=s'              => sub ($version) { $skip{$_} = 1 for split /,/, $version },
    'resume'              => \$opt_resume,
    'output|o=s'          => \$opt_output,
    'help|h'              => sub { pod2usage({ -verbose => 2}) },
    'branch|b=s'          => \$opt_branch,
    'version'             => sub {
      say 'cpangitify version ', ($App::cpangitify::VERSION // 'dev');
      exit 1;
    },
  ) || pod2usage(1);

  local $trace = $opt_trace;

  my @names = @ARGV;
  s/::/-/g for @names;
  my %names = map { $_ => 1 } @names;
  my $name = $names[0];

  pod2usage(1) unless $name;

  my $dest = $opt_output ? dir($opt_output)->absolute : dir()->absolute->subdir($name);

  if(-e $dest && ! $opt_resume)
  {
    say "already exists: $dest";
    say "you may be able to update with the --resume option";
    say "but any local changes to your repository will be overwritten by upstream";
    return 2;
  }

  say "creating/updating index...";
  my $history = CPAN::ReleaseHistory->new(
    maybe url => $opt_backpan_index_url
  )->release_iterator;

  say "searching...";
  my @rel;
  while(my $release = $history->next_release)
  {
    next unless defined $release->distinfo->dist;
    next unless $names{$release->distinfo->dist};
    push @rel, $release;
  }

  if($@ || @rel == 0)
  {
    say "no releases found for $name";
    return 2;
  }

  say "mkdir $dest";
  $dest->mkpath(0,0700);

  my $git = Git::Wrapper->new($dest->stringify);

  if($opt_resume)
  {
    if($git->status->is_dirty)
    {
      die "the appear to be uncommited changes";
    }
    $skip{$_} = 1 for $git->tag;
  }
  else
  {
    $git->init;
    $git->checkout( -b => $opt_branch );
  }

  foreach my $rel (@rel)
  {
    my $path    = $rel->path;
    my $version = $rel->distinfo->version;
    my $time    = $rel->timestamp;
    my $cpanid  = $rel->distinfo->cpanid;

    say "$path [ $version ]";

    if($skip{$version})
    {
      say "skipping ...";
      next;
    }

    my $tmp = dir( tempdir( CLEANUP => 1 ) );

    local $CWD = $tmp->stringify;

    my $uri = URI->new(join('/', $opt_backpan_url, $path));
    say "fetch ... $uri";
    my $res = $ua->get($uri);
    unless($res->{success})
    {
      say "error fetching $uri";
      say $res->{reason};
      return 2;
    }

    say "unpack... @{[ basename $uri->path ]}";
    my $extract = Archive::Libarchive::Extract->new(
      memory => \$res->{content},
      entry => sub ($e) {
        say "- extract @{[ $e->pathname ]}" if $trace;
        1;
      },
    );
    $extract->extract( to => $CWD );

    my $source = do {
      my @children = map { $_->absolute } dir()->children;
      if(@children != 1)
      {
        say "archive doesn't contain exactly one child: @children";
      }

      $CWD = $children[0]->stringify;
      $children[0];
    };

    say "merge...";

    foreach my $child ($dest->children)
    {
      next if $child->basename eq '.git';
      _rm_rf($child);
    }

    foreach my $child ($source->children)
    {
      next if $child->basename eq '.git';
      if(-d  $child)
      {
        rcopy($child, $dest->subdir($child->basename)) || die "unable to copy $child $!";
      }
      else
      {
        rcopy($child, $dest->file($child->basename)) || die "unable to copy $child $!";
      }
    }

    say "commit and tag...";
    $git->add('.');
    $git->add('-u');
    $git->commit({
      message       => "version $version",
      date          => "$time +0000",
      author        => author($cpanid),
      'allow-empty' => 1,
    });
    eval { $git->tag($version) };
    warn $@ if $@;
  }

  return 0;
}

1;
